pkg/controller/v1alpha2/applicationconfiguration/component.go

```diff
@@ -100,18 +104,25 @@ func (c *ComponentHandler) IsRevisionDiff(mt metav1.Object, curComp *v1alpha2.Co
 	if curComp.Status.LatestRevision == nil {
 		return true, 0
 	}
-	oldRev, err := c.AppsClient.ControllerRevisions(mt.GetNamespace()).Get(context.Background(), curComp.Status.LatestRevision.Name, metav1.GetOptions{})
-	if err != nil {
+
+	// client in controller-runtime will use infoermer cache
+	// use client will be more efficient
+	oldRev := &appsv1.ControllerRevision{}
+	if err := c.Client.Get(context.TODO(), client.ObjectKey{Namespace: mt.GetNamespace(), Name: curComp.Status.LatestRevision.Name}, oldRev); err != nil {
 		c.Logger.Info(fmt.Sprintf("get old controllerRevision %s error %v, will create new revision", curComp.Status.LatestRevision.Name, err), "componentName", mt.GetName())
 		return true, curComp.Status.LatestRevision.Revision
 	}
+	if oldRev.Name == "" {
+		c.Logger.Info(fmt.Sprintf("Not found controllerRevision %s", curComp.Status.LatestRevision.Name), "componentName", mt.GetName())
+		return true, curComp.Status.LatestRevision.Revision
+	}
 	oldComp, err := UnpackRevisionData(oldRev)
 	if err != nil {
 		c.Logger.Info(fmt.Sprintf("Unmarshal old controllerRevision %s error %v, will create new revision", curComp.Status.LatestRevision.Name, err), "componentName", mt.GetName())
 		return true, oldRev.Revision
 	}
 
-	if reflect.DeepEqual(curComp.Spec, oldComp.Spec) {
+	if curComp.Spec.Equal(&oldComp.Spec) {
```

Comment:
Why `reflect.DeepEqual` don't work?

Reply:
revision.Data read from etcd is  string, for example
`runtime.RawExtension{Object: runtime.Object}` save to etcd
and read from etcd will not decode Object, just be 
`runtime.RawExtension{Raw:[]byte}`
```

Reply:
that's should be logic contained in refactor client.

Reply:
`runtime.RawExtension` has no functions that unMarshal the raw data, and the Object is an interface not a struct, I'm afraid there're no way to undecode the raw data currently.

Reply:
I suggest we raise another PR to discuss it

Reply:
👌